whelktool/scripts/cleanups/2021/04/lxl-3547-link-expressionOf.groovy

```diff
@@ -355,6 +447,170 @@ private String titleStr(Map thing) {
     return props.grep().join(' ')
 }
 
+// Handle e.g. { "@type": "Language", "label": ["English & Tamil."] }
+private List mapBlankLanguages(List languages, List whichLanguageVersion = []) {
+    if (languages.size() == 1 && languages[0].label) {
+        String label = asList(languages[0].label).first().toLowerCase()
+        // to be able to map e.g. "Greek" to modern or classical Greek (LanguageLinker will use an existing linked sibling to decide when ambiguous)
+        List copy = new ArrayList( 
+                ambiguousLangNames.any{ label.contains(it) } 
+                    ? languages + whichLanguageVersion 
+                    : languages
+        )
+        Map m = ['l': copy]
+        languageLinker.linkLanguages(m, 'l')
+        return m['l']
+    }
+    
+    return languages
+}
+
+// mainTitle: "Haggada. Jiddisch & Hebreiska" --> mainTitle: Haggada, language: [[@id:https://id.kb.se/language/yid], [@id:https://id.kb.se/language/heb]]
+boolean moveLanguagesFromTitle(String id, Map work, List whichLanguageVersion = []) {
+    boolean changed = false
+    (asList(work['hasTitle'])?.first().mainTitle =~ /^(?<title>.*)\.(?<languages>[^.]+)\.?\s*$/).with {
+        if (matches()) {
+            String title = group('title')
+            List languages = group('languages')
+                    .split(",| & | och | and ")
+                    .collect{ Unicode.trimNoise(it.toLowerCase()) }
+
+            if (!languages || !languages.every{ it in languageNames }) {
+                return false
+            }
+            
+            def l = languages.collect{ ['@type': 'Language', 'label': it] }
+            if (languages.any{ it in ambiguousLangNames }) {
+                l += whichLanguageVersion
+            }
+
+            Map m = ['l': l]
+            languageLinker.linkLanguages(m, 'l')
+            l = m.l
+            if (l.size() == languages.size() && l.every {it['@id'] }) {
+                languageFromTitle.println("$id ${work.hasTitle.mainTitle} --> $title $l")
+                asList(work['hasTitle'])?.first().mainTitle = title
+                work.language = l
```

Comment:
Since expressionOf.language is overwritten, are we sure it never contains any language different to those found in expressionOf.hasTitle.mainTitle?

Reply:
Good catch!.
I meant to pass languages from expressionOf.hasTitle.mainTitle + expressionOf.language to `languageLinker` but accidentally removed the expressionOf.language part at some point